resources/js/callbacks.js

```diff
@@ -19,20 +19,79 @@ Vue.prototype.transformAmCustomFormResponse = async function (response) {
 
 Vue.prototype.amFormToVariables = function (form) {
     let variables = {form_data: {form_id: form.form_id}}
-    let replaceVariables = value => value
-        // Customer Variables
-        .replaceAll('{firstname}', window.app?.user?.firstname ?? '')
-        .replaceAll('{lastname}', window.app?.user?.lastname ?? '')
-        .replaceAll('{email}', window.app?.user?.email ?? '')
-        .replaceAll('{company}', window.app?.user?.addresses[0]?.company ?? '')
-        .replaceAll('{telephone}', window.app?.user?.addresses[0]?.telephone ?? '')
-        .replaceAll('{street}', window.app?.user?.addresses[0]?.street?.join(' ') ?? '')
-        .replaceAll('{city}', window.app?.user?.addresses[0]?.city ?? '')
-        .replaceAll('{region}', window.app?.user?.addresses[0]?.region?.region ?? '')
-        .replaceAll('{postcode}', window.app?.user?.addresses[0]?.postcode ?? '')
-        // General Variables
-        .replaceAll('{url}', window.location.href)
-        .replaceAll('{product_url}', window.location.href);
+
+    let navigate = (pointer, part) => {
+        if(!pointer) {
+            return null
+        }
+
+        if(part.startsWith('custom:')) {
+            return pointer.custom_attributes?.find(attribute => attribute.attribute_code === part.substring(7)).value
+        }
+        switch(part) {
+            case 'address':
+                return pointer.addresses?.[0]
+            case 'billing':
+            case 'shipping':
+                let id = pointer['default_'+part]
+                return pointer.addresses?.find(address => address.id == id)
+            case 'street':
+                return pointer.street?.join(' ')
+
+            default:
+                return pointer[part]
+        }
+    }
+
+    let getUserVariable = (parts) => {
+        let pointer = window.app?.user
+        if(!pointer) {
+            return ''
+        }
+
+        // Iterate through the parts to navigate the user object
+        for(const part of parts) {
+            let fallback = navigate(pointer?.addresses?.[0], part)
+            pointer = navigate(pointer, part)
+
+            // Return empty string when not defined
+            if(!pointer) {
+                if(fallback) {
+                    return fallback
+                }
+                return ''
+            }
+        }
+        return pointer
```

Comment:
I think it's best to move the fallback to the place where the pointer does not exist, so we're not needlessly traversing the object every time the fallback isn't needed
```suggestion
        for(const part of parts) {
            pointer = navigate(pointer, part)

            // Return empty string when not defined
            if(!pointer) {
                let fallback = navigate(pointer?.addresses?.[0], part)
                if(fallback) {
                    return fallback
                }
                return ''
            }
        }
        return pointer
```
or
```suggestion
        for(const part of parts) {
            pointer = navigate(pointer, part)

            if(pointer) {
                continue;
            }
            
            let fallback = navigate(pointer?.addresses?.[0], part)
            if(fallback) {
                return fallback
            }
            
            return ''
        }
        return pointer
```

Reply:
Problem there is that by the time we get to a null pointer, we can't traverse backwards and get the addresses. I guess I could just go and reset the pointer to `this.$root.user?.addresses?.[0]` to save on needlessly traversing every time?